soc/xtensa/intel_adsp/ace/power.c

```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 	z_xt_ints_on(core_desc[cpu].intenable);
 }
 
+#if CONFIG_ADSP_DYNAMIC_CLOCK_GATING
+
+atomic_t clock_gating_lock;
+
+void adsp_clock_gating_lock_get(void)
+{
+	if (atomic_get(&clock_gating_lock) == 0)
+		DSPCS.bootctl[arch_proc_id()].bctl |= DSPBR_BCTL_WAITIPCG;
+
+	atomic_inc(&clock_gating_lock);
+}
+
+void adsp_clock_gating_lock_put(void)
+{
+	atomic_t cnt = atomic_dec(&clock_gating_lock);
+
+	ARG_UNUSED(cnt);
+
+	__ASSERT(cnt >= 1, "Unbalanced clock gating lock get/put");
```

Comment:
Isn't this wrong. If we have just a single get and single put, "cnt" will be 0 zero and assert won't hold.